snow/sea ice shortwave radiative transfer harmonization

chengdang opened the pull request:
This merge adds an optional subroutine to the snow-covered sea ice shortwave radiative transfer calculations that allows the snow to be treated the same to the land snow model (SNICAR). The same subroutine also improves the shortwave calculation of near-IR band for snow-covered sea ice with new parameterizations.

akturner commented:
@chengdang: Please merge the latest version of MPAS-Seaice into your feature branch.

akturner commented on default_inputs/namelist.seaice:
```diff
@@ -0,0 +1,488 @@
+&seaice_model
```
This file should not have been added to the repository. Please remove it.

akturner commented on default_inputs/streams.seaice:
```diff
@@ -0,0 +1,420 @@
+<streams>
```
This file should not have been added to the repository. Please remove it.

akturner commented on namelist.seaice:
```diff
@@ -0,0 +1,489 @@
+&seaice_model
```
This file should not have been added to the repository. Please remove it.

akturner commented on src/rundir:
```diff
@@ -0,0 +1,489 @@
+&seaice_model
```
This file should not have been added to the repository. Please remove it.

akturner commented on streams.seaice:
```diff
@@ -0,0 +1,459 @@
+<streams>
```
This file should not have been added to the repository. Please remove it.

chengdang commented:
@akturner I removed the files you mentioned and snicar_optics_5bnd_snow_and_aerosols.nc, and updated namelist.seaice and streams.seaice under directory:  testing_and_setup/seaice/configurations/standard_physics/

chengdang commented on testing_and_setup/seaice/configurations/standard_physics/namelist.seaice:
```diff
@@ -3,7 +3,7 @@
     config_calendar_type = 'gregorian_noleap'
     config_start_time = '2000-01-01_00:00:00'
     config_stop_time = 'none'
-    config_run_duration = '00-00-01_00:00:00'
```
Okay, this change was made for model simulation. Revised

chengdang commented on testing_and_setup/seaice/configurations/standard_physics/namelist.seaice:
```diff
@@ -483,3 +483,7 @@
     config_AM_timeSeriesStatsCustom_reset_intervals = '00-00-07_00:00:00'
     config_AM_timeSeriesStatsCustom_backward_output_offset = '00-00-01_00:00:00'
 /
+&snicar_variables
```
done

chengdang commented on testing_and_setup/seaice/configurations/standard_physics/namelist.seaice:
```diff
@@ -483,3 +483,7 @@
     config_AM_timeSeriesStatsCustom_backward_output_offset = '00-00-01_00:00:00'
 /
+&snicar_variables
+    config_turn_snicar_system_on = true
```
done

chengdang commented on testing_and_setup/seaice/configurations/standard_physics/streams.seaice:
```diff
@@ -9,26 +9,107 @@
                   filename_interval="none"
```
I don't recall making any changes, perhaps an extra space? or indent?

chengdang replied:
I did while these two files were identical at line 9 under vimdiff. I rewrote updated streams.seaice using what's on master branch, please check if this is fixed.

chengdang commented on src/core_seaice/Registry.xml:
```diff
@@ -342,6 +353,7 @@
 			possible_values="true or false"
 		/>
 	</nml_record>
+
```
done

chengdang commented on src/core_seaice/Registry.xml:
```diff
@@ -1640,6 +1652,15 @@
 	</nml_record>
 
+	<!-- snicar variable namelist record -->
+	<nml_record name="snicar_variables" in_defaults="true">
```
done

chengdang commented on src/core_seaice/Registry.xml:
```diff
@@ -1640,6 +1652,15 @@
 
+	<!-- snicar variable namelist record -->
+	<nml_record name="snicar_variables" in_defaults="true">
+		 <nml_option name="config_turn_snicar_system_on" type="logical" default_value="false" units="unitless"
```
done

chengdang commented on src/core_seaice/Registry.xml:
```diff
@@ -1897,6 +1918,35 @@
+		        	immutable="true">
+                
+                    <!-- snow grain optical properties-->
+            		<var name="iceAsymmtryParameterDirect"/>
```
done

chengdang commented on src/core_seaice/Registry.xml:
```diff
@@ -2166,6 +2216,10 @@
 		<package name="pkgForcing" description="On if the framework forcing is being used"/>
 
+        <!-- Snicar -->
+        <package name="pkgSnicarSystem" description="add snicar 5 bands snow SSP in CICE"/>
```
done

chengdang commented on src/core_seaice/Registry.xml:
```diff
@@ -3748,6 +3802,36 @@
 
+	<!-- Snicar 5-band optical property varables -->
+	<var_struct name="snicar_variables" time_levs="1" packages="pkgSnicarSystem">
+        	<var name="iceAsymmtryParameterDirect"		type="real"	 dimensions="nSpectralIntervalsSNICAR nSnowGrainRadiusSNICAR" 	 name_in_code="iceAsymmtryParameterDirect"/>
```
done

chengdang commented on src/core_seaice/Registry.xml:
```diff
@@ -3748,6 +3802,36 @@
+        
+        	<var name="iceMassExtinctionCrossSectionDiffuse"	type="real" 	dimensions="nSpectralIntervalsSNICAR nSnowGrainRadiusSNICAR"	name_in_code="iceMassExtinctionCrossSectionDiffuse"/>
+            
+            <var name="aerosolMassExtinctionCrossSection_5bd"    type="real"    dimensions="nSpectralIntervalsSNICAR maxAerosolType"    name_in_code="aerosolMassExtinctionCrossSection_5bd"/>
```
fixed

chengdang commented on src/core_seaice/Registry.xml:
```diff
@@ -3748,6 +3802,36 @@
+            
+            <var name="aerosolMassExtinctionCrossSection_5bd"    type="real"    dimensions="nSpectralIntervalsSNICAR maxAerosolType"    name_in_code="aerosolMassExtinctionCrossSection_5bd"/>
+            
+            <var name="aerosolSingleScatterAlbedo_5bd"        type="real"    dimensions="nSpectralIntervalsSNICAR maxAerosolType"    name_in_code="aerosolSingleScatterAlbedo_5bd"/>
```
done

chengdang commented on src/core_seaice/column/ice_colpkg.F90:
```diff
@@ -2908,7 +2945,21 @@ subroutine colpkg_step_radiation (dt,       ncat,      &
                           l_print_point,                &
-                          linitonly)
+                          linitonly,                    &
+                          snicar_system,         &
```
done

chengdang commented on src/core_seaice/model_forward/mpas_seaice_core_interface.F:
```diff
@@ -578,11 +582,23 @@ subroutine setup_packages_other(configPool, packagePool, ierr)!{{{
+        pkgSnicarSystemActive = .true.
      endif
 
+!    call MPAS_pool_get_package(packagePool, "pkgSnicarSystem", pkgSnicarSystem)
```
removed

chengdang commented on src/core_seaice/Registry.xml:
```diff
@@ -3748,6 +3802,36 @@
 	</var_struct>
 
+	<!-- Snicar 5-band optical property varables -->
+	<var_struct name="snicar_variables" time_levs="1" packages="pkgSnicarSystem">
```
renamed

chengdang commented on src/core_seaice/shared/mpas_seaice_column.F:
```diff
@@ -2199,7 +2202,19 @@ subroutine column_radiation(domain, clock, lInitialization)
          albedoIRDirectCategory, &
          albedoIRDiffuseCategory, &
-         snowFractionCategory
+         snowFractionCategory,&
```
corrected

chengdang commented on src/core_seaice/shared/mpas_seaice_column.F:
```diff
@@ -2465,7 +2505,21 @@ subroutine column_radiation(domain, clock, lInitialization)
                pondLidMeltFluxFraction(:,iCell), &
                .false., &
-               lInitialization)
+               lInitialization,          &
```
fixed

chengdang commented on src/core_seaice/shared/mpas_seaice_column.F:
```diff
@@ -2619,6 +2673,7 @@ subroutine column_ridging(domain, ierr)
        call MPAS_pool_get_subpool(block % structs, "initial", initial)
        call MPAS_pool_get_subpool(block % structs, "velocity_solver", velocity_solver)
 
+
```
fixed

chengdang commented on src/core_seaice/shared/mpas_seaice_column.F:
```diff
@@ -9158,6 +9215,7 @@ subroutine init_column_package_configs(domain)
     call MPAS_pool_get_config(domain % configs, "config_scales_absorption_phaeocystis", config_scales_absorption_phaeocystis)
     call MPAS_pool_get_config(domain % configs, "config_ratio_C_to_N_proteins", config_ratio_C_to_N_proteins)
 
+
```
removed

chengdang commented:
Adrian, thanks for checking the code. Modified files have been pushed. Please check if there are more changes need to be made.

chengdang commented on src/core_seaice/column/ice_shortwave.F90:
```diff
@@ -1,30 +1,31 @@
+
```
I removed/added spaces to match the ice_shortwave.F90 on the master branch. Please check the updated code.

akturner commented on testing_and_setup/seaice/configurations/standard_physics/namelist.seaice:
```diff
@@ -256,6 +256,7 @@
 &shortwave
     config_shortwave_type = 'dEdd'
     config_albedo_type = 'ccsm3'
+    config_use_snicar = true
```
Why is this set to true?

chengdang replied:
It is set to false now

akturner commented on src/core_seaice/shared/mpas_seaice_column.F:
```diff
@@ -2199,7 +2202,19 @@ subroutine column_radiation(domain, clock, lInitialization)
          albedoIRDiffuseCategory, &
-         snowFractionCategory
+         snowFractionCategory,&
+         iceAsymmtryParameterDirect,     &
```
Fix this formatting: extra spaces

chengdang replied:
fixed

akturner commented on src/core_seaice/shared/mpas_seaice_column.F:
```diff
@@ -2219,7 +2234,9 @@ subroutine column_radiation(domain, clock, lInitialization)
          brineFraction, &
          modalBCabsorptionParameter, &
-         bioTracerShortwave
+         bioTracerShortwave,  &
```
formatting

chengdang replied:
fixed

akturner commented on src/core_seaice/column/ice_shortwave.F90:
```diff
@@ -765,6 +780,27 @@ subroutine run_dEdd(dt,       tr_aero,   &
+         ext_cff_mss_ice_dfs    ! snow mass extinction cross section (m2/kg)
+
+      real (kind=dbl_kind), dimension(:,:), intent(in) :: &
+        kaer_tab_5bd, & ! aerosol mass extinction cross section (m2/kg)
```
Be consistent with indenting

chengdang replied:
fixed

akturner commented on src/core_seaice/column/ice_shortwave.F90:
```diff
@@ -1189,6 +1258,11 @@ subroutine shortwave_dEdd  (n_aero,   n_zaero,     &
 
+      logical (kind=log_kind), intent(in) :: &
+         use_snicar    ! if true, use 5-band snicar IOPs for
+                           ! shortwave radiative calculation of
```
Consistent formatting

chengdang replied:
fixed comment format

akturner commented on src/core_seaice/column/ice_shortwave.F90:
```diff
@@ -1236,6 +1310,29 @@ subroutine shortwave_dEdd  (n_aero,   n_zaero,     &
+          ext_cff_mss_ice_dfs    ! snow mass extinction cross section (m2/kg)
+
+      real (kind=dbl_kind), dimension(:,:), intent(in) :: &
+         kaer_tab_5bd, & ! aerosol mass extinction cross section (m2/kg)
```
consistent indenting: keep original style

chengdang replied:
fixed

akturner commented on src/core_seaice/column/ice_shortwave.F90:
```diff
@@ -1339,7 +1437,42 @@ subroutine shortwave_dEdd  (n_aero,   n_zaero,     &
                srftyp = 1
-               call compute_dEdd(nilyr,       nslyr,   klev,   klevp,   & 
+
+              if (use_snicar) then ! use 5-band snicar IOPs for snow
```
formattign

chengdang replied:
fixed

akturner commented on src/core_seaice/column/ice_shortwave.F90:
```diff
@@ -1339,7 +1437,42 @@ subroutine shortwave_dEdd  (n_aero,   n_zaero,     &
+
+              if (use_snicar) then ! use 5-band snicar IOPs for snow
+
+                call compute_dEdd_5bd(nilyr, nslyr, klev, klevp,      &
```
formatting

chengdang replied:
fixed

akturner commented on src/core_seaice/column/ice_shortwave.F90:
```diff
@@ -1339,7 +1437,42 @@ subroutine shortwave_dEdd  (n_aero,   n_zaero,     &
+                        fswsfc,    fswint,                                &
+                        fswthru,   Sswabs,                                &
+                        Iswabs,    fswpenl,                               &
+                        asm_prm_ice_drc, asm_prm_ice_dfs, &
```
formatting

chengdang replied:
fixed

akturner commented on src/core_seaice/shared/mpas_seaice_column.F:
```diff
@@ -2353,6 +2371,21 @@ subroutine column_radiation(domain, clock, lInitialization)
        call MPAS_pool_get_array(biogeochemistry, "verticalGrid", verticalGrid)
 
+       ! snicar 5-band snow IOPs
+       call MPAS_pool_get_array(snicar_variables, "iceAsymmtryParameterDirect",iceAsymmtryParameterDirect)
```
formatting

akturner commented on src/core_seaice/shared/mpas_seaice_column.F:
```diff
@@ -2353,6 +2371,21 @@ subroutine column_radiation(domain, clock, lInitialization)
+       call MPAS_pool_get_array(snicar_variables, "aerosolSingleScatterAlbedo5band", aerosolSingleScatterAlbedo5band)
+       call MPAS_pool_get_array(snicar_variables,"aerosolAsymmetryParameter5band", aerosolAsymmetryParameter5band)
+       call MPAS_pool_get_array(snicar_variables, "modalMassExtinctionCrossSection5band", modalMassExtinctionCrossSection5band)
+       call MPAS_pool_get_array(snicar_variables,"modalSingleScatterAlbedo5band", modalSingleScatterAlbedo5band)
```
formatting

akturner commented on src/core_seaice/shared/mpas_seaice_column.F:
```diff
@@ -2353,6 +2371,21 @@ subroutine column_radiation(domain, clock, lInitialization)
+       call MPAS_pool_get_array(snicar_variables, "iceMassExtinctionCrossSectionDiffuse", iceMassExtinctionCrossSectionDiffuse)
+       call MPAS_pool_get_array(snicar_variables, "aerosolMassExtinctionCrossSection5band", aerosolMassExtinctionCrossSection5band)
+       call MPAS_pool_get_array(snicar_variables, "aerosolSingleScatterAlbedo5band", aerosolSingleScatterAlbedo5band)
+       call MPAS_pool_get_array(snicar_variables,"aerosolAsymmetryParameter5band", aerosolAsymmetryParameter5band)
```
formatting

akturner commented on src/core_seaice/shared/mpas_seaice_column.F:
```diff
@@ -2353,6 +2371,21 @@ subroutine column_radiation(domain, clock, lInitialization)
 
+       ! snicar 5-band snow IOPs
+       call MPAS_pool_get_array(snicar_variables, "iceAsymmtryParameterDirect",iceAsymmtryParameterDirect)
+       call MPAS_pool_get_array(snicar_variables, "iceAsymmtryParameterDiffuse",iceAsymmtryParameterDiffuse)
```
formatting

akturner commented:
In answer to your questions above:
1) Thanks!
2) Thanks for letting me know - I will look at this.
3) I will fix this, but generally this file shouldn't be used
4) You might want to add a directory to configurations, called "snicar_shortwave" with a streams file and namelist that tests your functionality.

chengdang commented:
Thanks! A directory (snicar_shortwave) with streams file and namelist is added to configurations for testing snicar related features.

akturner commented on testing_and_setup/seaice/configurations/standard_physics/streams.seaice:
```diff
@@ -9,6 +9,18 @@
                   filename_interval="none"
                   input_interval="initial_only" />
 
+<immutable_stream name="StandardAerosolsInput"
```
Yes, please remove these. Theyre not needed for the base physics simulation.

chengdang replied:
Sure. They are removed.

akturner commented on testing_and_setup/seaice/configurations/standard_physics/streams.seaice:
```diff
@@ -9,6 +9,12 @@
                   filename_interval="none"
                   input_interval="initial_only" />
 
+<immutable_stream name="SnicarInput"
```
This isnt needed in this configuration. Please remove it.

chengdang replied:
Removed.

akturner commented:
Please add a README file to the snicar_shortwave configuration describing its unique features

chengdang commented:
README file is added to configuration snicar_shortwave

akturner commented:
Few more modifications:
1) Add config_use_snicar = .false. to the appropriate place in the other namelists in the MPAS-Model/testing_and_setup/seaice/configurations/ directories
2) Change the streams file in MPAS-Model/testing_and_setup/seaice/configurations/snicar_shortwave so that "forcing/" appears in front of the aerosol files since this is where they will be pulled automatically during testing
3) Change the name of standard_optics_mpas_cice.nc in that streams file to standard_optics_mpas_seaice.nc since this is the name in the testing data set.

The snicar streams file should read:

<immutable_stream name="StandardAerosolsInput"
                  type="input"
                  filename_template="forcing/standard_optics_mpas_seaice.nc"
                  filename_interval="none"
                  input_interval="initial_only" />

<immutable_stream name="SnicarInput"
                  type="input"
                  filename_template="forcing/snicar_optics_5bnd_snow_and_aerosols.nc"
                  filename_interval="none"
                  input_interval="initial_only" />

chengdang commented:
Thanks Adrian. These are modified.

akturner commented on testing_and_setup/seaice/configurations/snicar_shortwave/streams.seaice:
```diff
@@ -0,0 +1,458 @@
+
+<immutable_stream name="SnicarInput"
+                  type="input"
+                  filename_template="forcing/snicar_optics_4bnd_snow_and_aerosols.nc"
```
Should not this be snicar_optics_5bnd_snow_and_aerosols.nc? Currently with the files I've been provided with this configuration fails because of this misnaming. Please test additions to PR.

chengdang replied:
You're right. Thanks for catching this typo. It is fixed.
